defender/client.go

```diff
@@ -0,0 +1,250 @@
+	if err := json.Unmarshal(body, &result); err != nil {
+		return fmt.Sprintf("No bearer token returned: %s", err)
+	}
+
+	if accessToken, ok := result["access_token"].(string); ok {
+		return accessToken
+	}
+
+	return ""
+
+}
+
+func (a *DefenderAdapter) fetchEvents(url string) {
+	defer a.wgSenders.Done()
+	defer a.conf.ClientOptions.DebugLog(fmt.Sprintf("fetching of %s events exiting", url))
+
+	lastEventId := ""
+	since := time.Now().Format("2006-01-02T15:04:05.000000Z")
+
+	for !a.doStop.WaitFor(30 * time.Second) {
+		// The makeOneRequest function handles error
+		// handling and fatal error handling.
+		items, newSince, eventId, _ := a.makeOneListRequest(url, since, lastEventId)
+		since = newSince
+		lastEventId = eventId
+		if items == nil {
+			continue
+		}
+
+		for _, item := range items {
+			msg := &protocol.DataMessage{
+				JsonPayload: item,
+				TimestampMs: uint64(time.Now().UnixNano() / int64(time.Millisecond)),
+			}
+			if err := a.uspClient.Ship(msg, 10*time.Second); err != nil {
+				if err == uspclient.ErrorBufferFull {
+					a.conf.ClientOptions.OnWarning("stream falling behind")
+					err = a.uspClient.Ship(msg, 1*time.Hour)
+				}
+				if err == nil {
+					continue
+				}
+				a.conf.ClientOptions.OnError(fmt.Errorf("Ship(): %v", err))
+				a.doStop.Set()
+				return
+			}
+		}
+	}
+}
+
+func (a *DefenderAdapter) makeOneListRequest(eventsUrl string, since string, lastEventId string) ([]map[string]interface{}, string, string, error) {
+
+	// Create query parameters
+	filter := "%24"
+	query := "%20ge%20"
+	date_filter := fmt.Sprintf("?%sfilter=createdDateTime%s%s", filter, query, strings.Replace(since, ":", "%3A", -1))
+
+	// Append query parameters to the URL
+	eventsUrl += date_filter
+
+	req, err := http.NewRequest("GET", eventsUrl, nil)
+	if err != nil {
+		a.conf.ClientOptions.OnError(fmt.Errorf("Error creating request: %s\n", err))
+		return nil, since, "", err
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", a.fetchToken()))
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		a.conf.ClientOptions.OnError(fmt.Errorf("Error making request: %s\n", err))
+		return nil, since, "", err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		a.conf.ClientOptions.OnError(fmt.Errorf("Error reading response: %s\n", err))
+		return nil, since, "", err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		a.conf.ClientOptions.OnError(fmt.Errorf("Error response: %s\n", body))
+		return nil, since, "", err
+	}
+
+	var data map[string]interface{}
+	err = json.Unmarshal(body, &data)
+	detections, _ := data["value"].([]interface{})
+	if err != nil {
+		a.conf.ClientOptions.OnError(fmt.Errorf("Error parsing JSON: %v", err))
+	}
+	items := detections
+
+	eventId := ""
+	lastDetectionTime := since
+	var alerts []map[string]interface{}
+	for _, detection := range items {
```

Comment:
All, looks good but just in case lets handle the cast in this block of code 

for _, detection := range items {
	detectMap, ok := detection.(map[string]interface{})
	if !ok {
		// Handle the error, e.g., log it and continue/return
	}

	id, ok := detectMap["id"].(string)
	if !ok {
		// Handle the error, e.g., log it and continue / return 
	}

	if id != lastEventId {
		createdDateTime, ok := detectMap["createdDateTime"].(string)
		if !ok {
			// Handle the error, e.g., log it and continue / return
		}

		lastDetectionTime = createdDateTime
		eventId = id
		alerts = append(alerts, detectMap)
	}
}